src/tool_operate.c

```diff
@@ -824,32 +873,92 @@ static CURLcode ipfs_url_rewrite(CURLU *uh, const char *protocol, char **url,
     }
   }
   else {
+    /* this is ensured to end in a trailing / within ipfs_gateway() */
     gateway = ipfs_gateway();
     if(!gateway) {
       result = CURLE_FILE_COULDNT_READ_FILE;
       goto clean;
     }
 
-    if(curl_url_set(ipfsurl, CURLUPART_URL, gateway, CURLU_GUESS_SCHEME
-                    | CURLU_NON_SUPPORT_SCHEME) != CURLUE_OK) {
+    if(curl_url_set(gatewaysurl, CURLUPART_URL, gateway, 0) != CURLUE_OK) {
+      result = CURLE_URL_MALFORMAT;
       goto clean;
     }
   }
 
-  pathbuffer = aprintf("%s/%s", protocol, cid);
+  /* check for unsupported gateway parts */
+  if(curl_url_get(gatewaysurl, CURLUPART_QUERY, &gatewayquery, 0)
+                  != CURLUE_NO_QUERY) {
+    result = CURLE_URL_MALFORMAT;
+    goto clean;
+  }
+
+  /* get gateway parts */
+  if(curl_url_get(gatewaysurl, CURLUPART_HOST,
+                  &gatewayhost, CURLU_URLDECODE) != CURLUE_OK) {
+    goto clean;
+  }
+
+  if(curl_url_get(gatewaysurl, CURLUPART_SCHEME,
+                  &gatewayscheme, CURLU_URLDECODE) != CURLUE_OK) {
+    goto clean;
+  }
+
+  curl_url_get(gatewaysurl, CURLUPART_PORT, &gatewayport, CURLU_URLDECODE);
+  curl_url_get(gatewaysurl, CURLUPART_PATH, &gatewaypath, CURLU_URLDECODE);
+
+  /* get the path from user input */
+  if(curl_url_get(uh, CURLUPART_PATH,
+                  &inputpath, CURLU_URLDECODE) != CURLUE_OK) {
+    inputpath = strdup("");
+    if(!inputpath)
+      goto clean;
+  }
+
+  /* set gateway parts in input url */
+  if(curl_url_set(uh, CURLUPART_SCHEME,
+                  gatewayscheme, CURLU_URLENCODE) != CURLUE_OK) {
+    goto clean;
+  }
+
+  if(curl_url_set(uh, CURLUPART_HOST,
+                  gatewayhost, CURLU_URLENCODE) != CURLUE_OK) {
+    goto clean;
+  }
+
+  if(curl_url_set(uh, CURLUPART_PORT,
+                  gatewayport, CURLU_URLENCODE) != CURLUE_OK) {
+    goto clean;
+  }
+
+  /* if the input path is just a slash, clear it */
+  if(inputpath && *inputpath && strlen(inputpath) == 1) {
+    if(*inputpath == '/') {
+      Curl_safefree(inputpath);
+      inputpath = strdup("");
+      if(!inputpath)
+        goto clean;
```

Comment:
Since the memory is already allocated and a single byte that you want to not be there. Maybe just overwrite it?

```suggestion
      *inputpath = '\0';
```